### PR TITLE
vo_gpu: improve HDR tone mapping (several changes)

### PR DESCRIPTION
**Pull request by haasn**
Ideally, you shouldn't need enter any text here, and your commit messages should
explain your changes sufficiently (especially why they are needed). Read
https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md for coding
style and development conventions. Remove this text block, but if you haven't
agreed to it before, leave the following sentence in place:

I agree that my changes can be relicensed to LGPL 2.1 or later.

### CODE REVIEW

**Comment by haasn** [comment]
Open questions:

- Is `--tone-mapping-max-boost` useful? Does it help on dark scenes? Or is it more annoying?
- Is `--tone-mapping-target-avg` useful? Or is the built-in default of 0.25 good enough?
- Does commit 55fb292 help or not?
- Are the default values of `--hdr-scene-threshold-low/high` good, or do they need to be reduced?

**Comment by haasn** [comment]
Cleaned up, squashed, rebased. Thanks to the help from @HyerrDoktyer, @aufkrawall, @Doofussy2 and @shinchiro (in issue #6405), we've gotten the defaults to a point where I feel confident merging this.

Resolves #6405 if merged.

**Comment by haasn** [comment]
Oh, I should also acknowledge @daddesio for coming up with the idea for the new scene change detection.

**Comment by haasn** [comment]
Fixed a few bugs and rebased; also switched to dB units for the scene change detection.

File: DOCS/man/options.rst [hunk 1]
```diff
@@ -5243,17 +5251,50 @@ The following video options are currently all specific to ``--vo=gpu`` and
...
+    it, while a value of 1.0 means that overly bright colors will tend towards
+    being completely white. (This is not always the case, especially not for
+    highlights that are near primary colors)
+
+    Values in between apply progressively more/less aggressive desaturation.
+    This setting helps prevent unnaturally oversaturated colors for
+    super-highlights, by (smoothly) turning them into less saturated (per
+    channel tone mapped) colors instead. This makes images feel more natural,
+    at the cost of chromatic distortions for out-of-range colors. The default
+    value of 0.5 provides a good balance. Setting this to 0.0 preserves the
```
Comment (MoSal) [inline DOCS/man/options.rst]: This seems to be a left-over. The right default (0.75) is stated at the beginning.

**Comment by MoSal** [comment]
Didn't play with all the options, but `tone-mapping-max-boost=6.5` works great for me. Thanks.

**Comment by HyerrDoktyer** [comment]
@MoSal 
> Didn't play with all the options, but `tone-mapping-max-boost=6.5` works great for me. Thanks.

I suggest being quite conservative with this value if you're watching real HDR films (not demos), I've noticed that it can, in some cases, cause a bit of overexposure. Personally I've found `tone-mapping-max-boost=1.5` or `tone-mapping-max-boost=2.0` to be a quite a reasonable configuration. Even with these (great) improvements mpv does seem to be slightly too dark, and I'd like to see that fixed on a more fundamental level rather than just throwing max-boost at it. However gathering valid samples has been a bit of a headache...

**Comment by MoSal** [comment]
@HyerrDoktyer 

I do indeed tend to favor over-exposure over under-exposure. A value between 3.0 and 5.0 might suit others better. Anything lower is not a significant improvement IMHO.

I already have
```
H add tone-mapping-max-boost +0.5
G add tone-mapping-max-boost -0.5
```

in `input.conf` in case I run into something that looks wrong.

**Comment by Doofussy2** [comment]
I've actually switched to using reinhard with peak detection and Max boost at 1.5. I find that gives me a better average of brightness and color. Titles like Interstellar are just on the edge of being too bright, but still within the limit. Chess also plays reasonably well. But if I compare that to using Mobius with target peak set to 1000, there is a big difference. I know there should be a difference, but this really shows just how big the light and dark difference, is. Some titles play well, others are clearly too dark. Any value less than 1000, causes a significant loss of detail.

**Comment by HyerrDoktyer** [comment]
@MoSal 
Fair enough, thanks for reminding me about the input.conf binds!


@Doofussy2 
Ha, funny you should say that! Just last night I was considering if it were possible to dim-down reinhard instead of attempting to dim down reinhard instead of attempting to raise the brightness with hable. The config you've suggested here is far, far too dim on my end (--target-peak also affects mpvs OSC (WHY?!?))

--target-peak=150 seems to give me a good result

Current defaults:
current_defaults

--tone-mapping=reinhard --target-prim=bt.2020 --target-peak=250:
target-peak150

--tone-mapping=reinhard --target-prim=bt.2020 --target-peak=1000:
target-peak1000

I'm somewhat concerned about highlight details on this, Interstellar is probably an ideal film to test that; I also have an interesting sample regarding "The First Man", the HDR master is incredible in comparison to the SDR (which once again, raises the question regarding the validity of directly comparing HDR films to their SDR counterparts in certain scenarios)

**Comment by Doofussy2** [comment]
Yeah, I just discovered a few titles that where too dark when using `--target-peak=1000`. Anything less than that and I get quite a lot detail degradation. However, I just tried using;

hdr-compute-peak=yes
tone-mapping=reinhard
target-prim=bt.2020
tone-mapping-param=0.25

And that gave me very good results. I recently was watching the new version of Lost in Space, which I had some problems with the darkness. Reinhard seems to negate the need for max boost, but can be a little too bright. I can't use mobius at all. It gives horrible brightness. Anyway, lowering the mapping param with reinhard, appears to work pretty well. My colors seem good, no issues with too much darkness, and just takes the edge off some of the really bright scenes.

Examples:

Chess (note the crease in 4th strip from the left. Details like that are being lost with other options)
mpv-shot0001


Lost in Space
mpv-shot0002


Interstellar
mpv-shot0003

Life of Pi
mpv-shot0004

**Comment by HyerrDoktyer** [comment]
If I'm understanding your question correctly, mpv defaults are hable with peak detection enabled
https://mpv.io/manual/master/#options-tone-mapping
https://mpv.io/manual/master/#options-hdr-compute-peak

**Comment by Doofussy2** [comment]
> If I'm understanding your question correctly, mpv defaults are hable with peak detection enabled

I incorrectly thought that haasn had changed that in these builds.

Juxtaposing hable and reinhard with mapping param 0.25, I think I prefer reinhard. It's a little brighter without losing too much detail, and the color looks just as good.

**Comment by MoSal** [comment]
[This sample](https://archive.org/download/unsorted_files/hdr_sample.mkv) (look at the water 1 second in) might be of interest.

`hable`, with any `tone-mapping-max-boost` value, handles this well.

**Comment by Doofussy2** [comment]
I just discovered that desaturation is what has been affecting the detail loss. Also, when using reinhard with peak detection, transition between light and dark, can be very noticeable. A good example of this is the [Samsung Wonderland](https://4kmedia.org/ultra-hd-hdr-samsung-4k-demo-wonderland/) video. At the beginning, the clouds are very light in color, and then you see peak detection pull it back in to what it should be. Disabling peak detection and setting `--target-peak=220` gives the correct color and brightness, without transition. And for the desaturation, I find that `--tone-mapping-desaturate=2.0` restores the detail, pretty well, without degradation in other scenes.

Present config:

hdr-compute-peak=no
tone-mapping=reinhard
target-prim=bt.2020
tone-mapping-param=0.35
target-peak=220
tone-mapping-desaturate=2.0

File: video/out/gpu/video_shaders.c [hunk 2]
```diff
@@ -567,123 +567,106 @@ static void pass_inverse_ootf(struct gl_shader_cache *sc, enum mp_csp_light ligh
...
-    GLSLF("  uint next = (frame_idx + 1) %% %d;\n", PEAK_DETECT_FRAMES+1);
-    GLSL(    total_max += cur_max - frame_max[next];)
-    GLSL(    total_avg += cur_avg - frame_avg[next];)
-    GLSL(    frame_max[next] = frame_avg[next] = 0;)
-
-    // Update the index and count
-    GLSL(    frame_idx = next;)
-    GLSLF("  frame_num = min(frame_num + 1, %d);\n", PEAK_DETECT_FRAMES);
+    GLSL(    vec2 cur = vec2(float(frame_sum) / float(num_wg), frame_max);)
+    GLSLF("  cur *= vec2(1.0/%f, 1.0/%f);\n", log_scale, sig_scale);
```
Comment (sfan5) [inline video/out/gpu/video_shaders.c]: is there any reason the `1.0/` was moved out of glsl on line 383 but not here?

Comment (haasn) [reply video/out/gpu/video_shaders.c]: I try to make my `%f` constants greater than 1. So for e.g. `log_scale` and `sig_scale` which are both pretty big numbers, I template it as `"1.0/%f", x`, but for the other number (the one I changed), the number `x` itself is <1.0 so I template it as `"%f", 1.0/x`.

Hope that makes sense.

**Comment by sfan5** [comment]
diff.pics has not started working yet, so I'll just dump the screenshots here:
"Samsung Wonderland Demo.ts" with `hdr-compute-peak=yes`, `profile=gpu-hq`
comparing with screenshots seen in #6072 the patch brings tonemapping closer to youtube's

<details>
<summary>without 6415</summary>



</details>
<details>
<summary>with 6415</summary>



</details>

**Comment by Doofussy2** [comment]
Let's get this merged to master :D

**Comment by jeeb** [comment]
Currently testing this (the code generally seems OK):

1. TravelXP sample seabed scene `--pause --start=60` (HLG)
   * [Before](https://user-images.githubusercontent.com/27182/52920761-030ea080-3319-11e9-9649-dbe7a2aa969c.png)
   * [After](https://user-images.githubusercontent.com/27182/52920798-5ed92980-3319-11e9-977f-1c682ca316d3.png)

**Pull request merged**